File: apps/api/src/schemas/rfid.py
"""Pydantic models for RFID management.

Defines request / response schemas for RFID authorization and access logs.
"""

from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class RFIDScanType(str, Enum):
    """Type of RFID scan event."""

    CHECKPOINT = "checkpoint"
    PICKUP = "pickup"
    DROPOFF = "dropoff"
    UNLOCK = "unlock"
    DENIED = "denied"


class RFIDTagStatus(str, Enum):
    """Current authorization status of an RFID tag."""

    ACTIVE = "active"
    REVOKED = "revoked"
    EXPIRED = "expired"
    UNKNOWN = "unknown"


class ContainerStatus(str, Enum):
    """Physical state of the delivery container lock."""

    LOCKED = "locked"
    UNLOCKED = "unlocked"
    ERROR = "error"


# ── Request schemas ──────────────────────────────────────────────────────


class RFIDAuthorizeRequest(BaseModel):
    """Payload for POST /rfid/authorize — simulate an RFID tap."""

    tag_id: str = Field(
        ...,
        min_length=4,
        max_length=32,
        description="RFID tag UID (e.g., 'RFID-A1B2C3')",
    )
    robot_id: str = Field(
        ...,
        min_length=1,
        description="ID of the robot whose container is being scanned",
    )
    delivery_id: Optional[str] = Field(
        None,
        description="Associated delivery (optional; can be auto-resolved)",
    )


class RFIDRegisterRequest(BaseModel):
    """Payload for registering a new RFID tag to a user."""

    tag_id: str = Field(..., min_length=4, max_length=32)
    user_name: str = Field(..., min_length=1, max_length=100)
    role: str = Field("recipient", pattern=r"^(operator|recipient|admin)$")


# ── Response schemas ─────────────────────────────────────────────────────


class RFIDAuthorizeResponse(BaseModel):
    """Result of an RFID authorization attempt."""

    authorized: bool
    tag_id: str
    robot_id: str
    delivery_id: Optional[str] = None
    container_status: ContainerStatus
    user_name: Optional[str] = None
    message: str
    timestamp: datetime


class RFIDLogEntry(BaseModel):
    """A single RFID access log record."""

    id: str
    tag_id: str
    robot_id: str
    delivery_id: Optional[str] = None
    scan_type: RFIDScanType
    authorized: bool
    user_name: Optional[str] = None
    location: Optional[str] = None
    message: str
    timestamp: datetime


class RFIDLogListResponse(BaseModel):
    """Paginated list of RFID access logs."""

    logs: list[RFIDLogEntry]
    total: int
    timestamp: datetime


class RFIDTagResponse(BaseModel):
    """Registered RFID tag info."""

    tag_id: str
    user_name: str
    role: str
    status: RFIDTagStatus
    registered_at: datetime
    last_used: Optional[datetime] = None

File: apps/web/lib/api/rfid.ts
/**
 * Stair-Doc RFID API client
 *
 * Typed fetch wrapper for RFID authorization, access logs, and tag management.
 */

import { OpenAPI } from "@/lib/api/client";

// ── Types matching backend Pydantic schemas ──────────────────────────────

export type RFIDScanType =
  | "checkpoint"
  | "pickup"
  | "dropoff"
  | "unlock"
  | "denied";

export type RFIDTagStatus = "active" | "revoked" | "expired" | "unknown";

export type ContainerStatus = "locked" | "unlocked" | "error";

export interface RFIDAuthorizeResponse {
  authorized: boolean;
  tag_id: string;
  robot_id: string;
  delivery_id: string | null;
  container_status: ContainerStatus;
  user_name: string | null;
  message: string;
  timestamp: string;
}

export interface RFIDLogEntry {
  id: string;
  tag_id: string;
  robot_id: string;
  delivery_id: string | null;
  scan_type: RFIDScanType;
  authorized: boolean;
  user_name: string | null;
  location: string | null;
  message: string;
  timestamp: string;
}

export interface RFIDLogListResponse {
  logs: RFIDLogEntry[];
  total: number;
  timestamp: string;
}

export interface RFIDTagResponse {
  tag_id: string;
  user_name: string;
  role: string;
  status: RFIDTagStatus;
  registered_at: string;
  last_used: string | null;
}

export interface RFIDAuthorizePayload {
  tag_id: string;
  robot_id: string;
  delivery_id?: string | null;
}

export interface RFIDRegisterPayload {
  tag_id: string;
  user_name: string;
  role?: string;
}

// ── API Response envelope ────────────────────────────────────────────────

interface ApiResponse<T> {
  success: boolean;
  data: T;
  message: string;
  timestamp: string;
}

// ── API Error ────────────────────────────────────────────────────────────

export class RFIDApiError extends Error {
  constructor(
    message: string,
    public statusCode: number,
    public body?: unknown,
  ) {
    super(message);
    this.name = "RFIDApiError";
  }
}

// ── Fetch helper ─────────────────────────────────────────────────────────

async function apiFetch<T>(path: string, init?: RequestInit): Promise<T> {
  const url = `${OpenAPI.BASE}${path}`;

  const response = await fetch(url, {
    ...init,
    headers: {
      "Content-Type": "application/json",
      ...init?.headers,
    },
  });

  if (!response.ok) {
    const body = await response.json().catch(() => null);
    throw new RFIDApiError(
      body?.detail ?? body?.message ?? `Request failed: ${response.status}`,
      response.status,
      body,
    );
  }

  const envelope: ApiResponse<T> = await response.json();

  if (!envelope.success) {
    throw new RFIDApiError(envelope.message, response.status);
  }

  return envelope.data;
}

// ── RFID API ─────────────────────────────────────────────────────────────

export interface FetchRFIDLogsParams {
  tag_id?: string;
  robot_id?: string;
  scan_type?: RFIDScanType;
  authorized?: boolean;
  limit?: number;
}

/**
 * Authorize an RFID scan (simulate container unlock).
 */
export async function authorizeRFID(
  payload: RFIDAuthorizePayload,
): Promise<RFIDAuthorizeResponse> {
  return apiFetch<RFIDAuthorizeResponse>("/api/v1/rfid/authorize", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

/**
 * Fetch RFID access logs with optional filters.
 */
export async function fetchRFIDLogs(
  params?: FetchRFIDLogsParams,
): Promise<RFIDLogListResponse> {
  const searchParams = new URLSearchParams();
  if (params?.tag_id) searchParams.set("tag_id", params.tag_id);
  if (params?.robot_id) searchParams.set("robot_id", params.robot_id);
  if (params?.scan_type) searchParams.set("scan_type", params.scan_type);
  if (params?.authorized !== undefined)
    searchParams.set("authorized", String(params.authorized));
  if (params?.limit) searchParams.set("limit", String(params.limit));

  const qs = searchParams.toString();
  return apiFetch<RFIDLogListResponse>(`/api/v1/rfid/logs${qs ? `?${qs}` : ""}`);
}

/**
 * List all registered RFID tags.
 */
export async function fetchRFIDTags(): Promise<RFIDTagResponse[]> {
  return apiFetch<RFIDTagResponse[]>("/api/v1/rfid/tags");
}

/**
 * Register a new RFID tag.
 */
export async function registerRFIDTag(
  payload: RFIDRegisterPayload,
): Promise<RFIDTagResponse> {
  return apiFetch<RFIDTagResponse>("/api/v1/rfid/tags", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}
